ompi/mca/coll/xhc/coll_xhc.c

```diff
@@ -9,609 +9,677 @@
-        ret = xhc_print_info(module, comm, data);
-        if(ret != OMPI_SUCCESS) {
-            RETURN_WITH_ERROR(return_code, ret, end);
+        if(err != OPAL_SUCCESS) {
+            opal_output_verbose(MCA_BASE_VERBOSE_WARN,
+                ompi_coll_base_framework.framework_output,
+                "coll:xhc: Warning: opal_hash_table_set_value_ptr() "
+                "returned %d", err);
         }
     }
 
-    // ----
+    // ---
 
-    module->data = data;
-    module->init = true;
+    err = xhc_comms_make(comm, module, config, data);
+    if(err != OMPI_SUCCESS) RETURN_WITH_ERROR(return_code, err, end);
 
-    end:
+    data->init = true;
 
-    xhc_module_install_fns(module, comm, xhc_fns);
+    if(mca_coll_xhc_component.print_info & (1 << colltype)) {
+        err = xhc_print_op_info(module, comm, colltype);
 
-    free(peer_cico_ds);
+        if(err != OMPI_SUCCESS) {
+            opal_output_verbose(MCA_BASE_VERBOSE_WARN,
+                ompi_coll_base_framework.framework_output,
+                "coll:xhc: Warning: xhc_print_op_info() "
+                "failed with error code %d", err);
+        }
 
-    if(return_code != 0) {
-        opal_show_help("help-coll-xhc.txt", "xhc-init-failed", true,
-            return_code, errno, strerror(errno));
+        if(mca_coll_xhc_component.print_info & XHC_PRINT_INFO_HIER_DOT) {
+            err = xhc_print_op_hierarchy_dot(module, comm, colltype);
 
-        xhc_fini(module);
+            if(err != OMPI_SUCCESS) {
+                opal_output_verbose(MCA_BASE_VERBOSE_WARN,
+                    ompi_coll_base_framework.framework_output,
+                    "coll:xhc: Warning: xhc_print_op_hierarchy_dot() "
+                    "failed with error code %d", err);
+            }
+        }
+    }
+
+    // ---
+
+    end:
+
+    if(return_code != OMPI_SUCCESS) {
+        opal_show_help("help-coll-xhc.txt", "xhc-op-init-failed", true,
+            xhc_colltype_to_str(colltype), return_code, errno, strerror(errno));
     }
 
     return return_code;
 }
 
-void mca_coll_xhc_fini(mca_coll_xhc_module_t *module) {
-    if(module->data) {
-        xhc_data_t *data = module->data;
+static int xhc_alloc_bcast_cico(xhc_module_t *module, ompi_communicator_t *comm) {
+    opal_shmem_ds_t *ds_list = NULL;
+    opal_shmem_ds_t cico_ds;
+    void *cico_buffer = NULL;
+    xhc_coll_fns_t xhc_fns;
 
-        if(data->comm_count >= 0) {
-            xhc_comms_destroy(data->comms, data->comm_count);
-        }
+    int err, return_code = OMPI_SUCCESS;
+
+    int comm_size = ompi_comm_size(comm);
+    int rank = ompi_comm_rank(comm);
+
+    size_t cico_size = module->op_config[XHC_BCAST].cico_max;
+
+    if(cico_size == 0)
+        return OMPI_SUCCESS;
+
+    xhc_module_set_coll_fns(comm, &module->prev_colls, &xhc_fns);
+
+    // --
 
-        free(data->comms);
-        free(data);
+    ds_list = malloc(comm_size * sizeof(opal_shmem_ds_t));
+    if(!ds_list) return OMPI_ERR_OUT_OF_RESOURCE;
```

Comment:
Minor: (sorry to bring up again an Open MPI coding guideline). Even single line blocks should be encapsulated in a pair of {}

Reply:
Thanks for taking the time to look into the PR. I'm less emotional about this one :-). Once again, vis-a-vis the safety aspect, I'd argue that GCC has `-Wmisleading-indentation` in `-Wall` for this purpose. I find it useful to not always encapsulate in `{}` especially for single-line ifs, because otherwise one can end up with a lot of extra lines, making functions longer and harder to follow. Of course now that I think of it, one could still have the braces on a single line (a bit unusual, but certainly doable, and perhaps better than all the extra lines).

I avoided converting these occurrences to make it easier to compare and keep track of changes between my development repo and the upstream version, to make sure I don't miss anything. But once again, if it must be, I can put in the work to do these conversions.